src/main/java/com/hashedin/amfi/NavService.java

```
@@ -0,0 +1,95 @@
+		this.amfiNavUrl = AMFI_NAV_URL;
+	}
+		
+	/**
+	 * Gets the NAV for a mutual fund for a given date
+	 * @param id the primary key for the mutual fund
+	 * @return NAV multiplied by 10^4
+	 */
+	public long getLatestNav(String id) {
+		return fetchNav(id);
+	}
+	
+	/*
+	 * Fetches NAV from AMFI's REST API
+	 * 
+	 * TODO: Performance Optimization 
+	 * This method is downloading a large file and only consuming a very small portion of the file
+	 * We should cache the response. Alternatively, we should find another REST API
+	 * 
+	 */
+	private long fetchNav(String id) {
+		/*
+		 * Raw Response is a very big file
+		 * Each line contains a mutual fund and it's details, including its NAV
+		 * Each line has tokens that are ; separated
+		 * 
+		 * Here is how it looks like:
+		 * 
+		 * Scheme Code;ISIN Div Payout/ ISIN Growth;ISIN Div Reinvestment;Scheme Name;Net Asset Value;Date
+		 * 
+		 * Open Ended Schemes(Debt Scheme - Banking and PSU Fund)
+		 * 
+		 * 
+		 * Aditya Birla Sun Life Mutual Fund
+		 * 119551;INF209KA12Z1;INF209KA13Z9;Aditya Birla Sun Life Banking & PSU Debt Fund  - Direct Plan-Dividend;159.817;15-Feb-2021
+		 * 119552;INF209K01YM2;-;Aditya Birla Sun Life Banking & PSU Debt Fund - Direct Plan-Monthly Dividend;114.3157;15-Feb-2021
+		 * 119553;INF209K01YO8;-;Aditya Birla Sun Life Banking & PSU Debt Fund - Direct Plan-Quarterly Dividend;112.2331;15-Feb-2021
+		 * 
+		 */
+		String rawResponse = this.restTemplate.getForObject(amfiNavUrl, String.class);
+		
+		String lines[] = rawResponse.split("\n");
+		for (String line : lines) {
+			if (line.isBlank()) {
+				continue;
+			}
+			if (line.startsWith(id)) {
+				String tokens[] = line.split(";");
+				if (tokens.length != 6) {
+					throw new IllegalStateException("AMFI response format has changed, expected exactly 6 tokens. \n" + line);
+				}
+				String navAsText = tokens[4];
+				return toLong(navAsText);
+			}
+		}
+		throw new IllegalArgumentException("Mutual Fund ID " + id + " not found");
```

We processed the entire file but could not find the NAV for our mutual fund. What do we do?

We have a few options - 
1. Return -1 to indicate we couldn't find the mutual fund. This is always a bad idea. Every code block that needs to process NAV will have to check `if (nav != -1)`. Bad idea.
2. Log an error. This is a better solution, but logging an error does not stop the flow. You still have to return a value, and you can't return -1 or null. 
3. You throw an exception, like we are doing. When you throw an exception, and if the caller does not catch it, ultimately spring will catch it for you. When it catches the exception, it will a) log the exception with stack trace as an error, and b) return a 500 error to the client. This is exactly what you want, and so IMO it is better than logging.
4.You can log an error and throw an exception. This just pollutes the logs. You will see two messages in your log files relating to the same problem. There is no guarantee that the messages will be one after the other. It just makes debugging more difficult.

---

Another imporant consideration - should we throw a custom exception, say something like `MutualFundNotFoundException`? 

Yes, throwing a custom exception, especially from a public method, has some merits. 

Notice that when Spring catches the exception, it returns a 500 error to the client. But if the mutual fund id is coming from the client, then it shouldn't be a 500, rather it should be a 400. 

We will handle the above in a subsequent pull request when we actually build the REST service.